lib/phlex/component.rb

```diff
@@ -1,27 +1,112 @@
 # frozen_string_literal: true
+require "digest"
+require "byebug"
 
 module Phlex
   class Component
     include Node, Context
 
+    class << self
+      def cache_key
+        @cache_key ||= {
+          template: checksum,
+          subcomponents: subcomponents.map(&:checksum),
+          ancestors: dependant_ancestor_checksums
+        }
+      end
+
+      def checksum
+        @checksum ||= Digest::MD5.hexdigest(file)
+      end
+
+      def subcomponents
+        @subcomponents ||= direct_subcomponents.reduce(Set.new) do |set, component|
+          set << component
+          component.subcomponents.each { set << _1 }
+          set
+        end
+      end
+
+      private
+
+      def dependant_ancestor_checksums
+        dependant_ancestors.map do |ancestor|
+          Digest::MD5.hexdigest(
+            File.read(
+              Module.const_source_location(ancestor.name).first
+            )
+          )
+        end
+      end
+
+      def dependant_ancestors
+        ancestors.map
+          .reject { [self, Object, BasicObject, Kernel].include? _1 }
+          .reject { _1.name.start_with? "Phlex::" }
+      end
+
+      def direct_subcomponents
+        file.scan(/component\s([A-Z]+[A-Za-z:]*)/).flatten.map { const_get(_1) }
```

Comment:
`const_get` here won't look up relative constants based on the namespace, so we need to write something else that will do that.